core/src/main/scala/org/apache/spark/resource/ResourceUtils.scala

```diff
@@ -70,6 +79,186 @@ private[spark] object ResourceUtils extends Logging {
+      if (resourcesFile.exists()) {
+        val allocated = {
+          val allocated = parseAllocatedFromJsonFile(resourcesFile.getPath)
+          allocated.map { allocation => allocation.id.resourceName -> allocation.addresses.toArray}
+        }.toMap
+        val newAllocated = {
+          allocated.map { case (rName, addresses) =>
+            val retained = addresses.diff(Option(toRelease.getOrElse(rName, null))
+              .map(_.addresses).getOrElse(Array.empty))
+            rName -> new ResourceInformation(rName, retained)
+          }.filter(_._2.addresses.nonEmpty)
+        }
+        if (newAllocated.nonEmpty) {
+          writeResourceAllocationJson(SPARK_WORKER_PREFIX, newAllocated, resourcesFile)
+        } else {
+          if (!resourcesFile.delete()) {
+            logWarning(s"Failed to delete $ALLOCATED_RESOURCES_FILE.")
+          }
+        }
+      }
+      releaseLock(lock)
+    }
+  }
+
+  private def acquireLock(): FileLock = {
+    val resourcesDir = getOrCreateResourcesDir()
+    val lockFile = new File(resourcesDir, RESOURCES_LOCK_FILE)
+    val lockFileChannel = new RandomAccessFile(lockFile, "rw").getChannel
+    var keepTry = true
+    var lock: FileLock = null
+    while(keepTry) {
+      try {
+        lock = lockFileChannel.lock()
+        logInfo(s"Acquired lock on $RESOURCES_LOCK_FILE.")
+        keepTry = false
+      } catch {
+        case e: OverlappingFileLockException =>
+          // This exception throws when we're in LocalSparkCluster mode. FileLock is designed
+          // to be used across JVMs, but our LocalSparkCluster is designed to launch multiple
+          // workers in the same JVM. As a result, when an worker in LocalSparkCluster try to
+          // acquire the lock on `resources.lock` which already locked by other worker, we'll
+          // hit this exception. So, we should manually control it.
+          keepTry = true
+          // there may be multiple workers race for the lock,
+          // so, sleep for a random time to avoid possible conflict
+          val duration = Random.nextInt(1000) + 1000
+          Thread.sleep(duration)
+      }
+    }
+    assert(lock != null, s"Acquired null lock on $RESOURCES_LOCK_FILE.")
+    lock
+  }
+
+  private def releaseLock(lock: FileLock): Unit = {
+    try {
+      lock.release()
+      lock.channel().close()
+      logInfo(s"Released lock on $RESOURCES_LOCK_FILE.")
+    } catch {
+      case e: Exception =>
+        logWarning(s"Error while releasing lock on $RESOURCES_LOCK_FILE.", e)
+    }
+  }
+
+  private def getOrCreateResourcesDir(): File = {
+    val sparkHome = if (Utils.isTesting) {
+      assert(sys.props.contains("spark.test.home"), "spark.test.home is not set!")
+      new File(sys.props("spark.test.home"))
+    } else {
+      new File(sys.env.getOrElse("SPARK_HOME", "."))
+    }
+    val resourceDir = new File(sparkHome, SPARK_RESOURCES_DIRECTORY)
```

Comment:
This really can be a problem. Make it configurable may be fine. If so, should we throw exception if user do not config it explicitly ?

Reply:
And we should avoid user to config it to a tmp directory, right ?

Reply:
If user doesn't config it and we leave it to default to sparkhome and they don't have permissions then it will fail anyway. I think this is similar to the SPARK_WORKER_DIR.
I don't think we need to check is user puts in tmp, if that is where they want it, they should realize consequences. We can document it.